### ISSUE TITLE
Unicode decode error when trying to read in ascii table

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#7047]
I thought that in Python 3, astropy.io.ascii worked with non-ASCII characters, but I am having issues with the following table:

https://raw.githubusercontent.com/mrwilson/strictly-come-dancing-results/master/results.csv

```python
In [4]: from astropy.table import Table

In [5]: t = Table.read('results.csv')
---------------------------------------------------------------------------
UnicodeEncodeError                        Traceback (most recent call last)
<ipython-input-5-aa54d4eed93f> in <module>()
----> 1 t = Table.read('results.csv')

~/Dropbox/Code/Astropy/astropy/astropy/table/table.py in read(cls, *args, **kwargs)
   2505         passed through to the underlying data reader (e.g. `~astropy.io.ascii.read`).
   2506         """
-> 2507         out = io_registry.read(cls, *args, **kwargs)
   2508         # For some readers (e.g., ascii.ecsv), the returned `out` class is not
   2509         # guaranteed to be the same as the desired output `cls`.  If so,

~/Dropbox/Code/Astropy/astropy/astropy/io/registry.py in read(cls, format, *args, **kwargs)
    515 
    516         reader = get_reader(format, cls)
--> 517         data = reader(*args, **kwargs)
    518 
    519         if not isinstance(data, cls):

~/Dropbox/Code/Astropy/astropy/astropy/io/ascii/connect.py in read_csv(filename, **kwargs)
     78     from .ui import read
     79     kwargs['format'] = 'csv'
---> 80     return read(filename, **kwargs)
     81 
     82 

~/Dropbox/Code/Astropy/astropy/astropy/io/ascii/ui.py in read(table, guess, **kwargs)
    366             fast_reader_rdr = get_reader(**fast_kwargs)
    367             try:
--> 368                 dat = fast_reader_rdr.read(table)
    369                 _read_trace.append({'kwargs': fast_kwargs,
    370                                     'Reader': fast_reader_rdr.__class__,

~/Dropbox/Code/Astropy/astropy/astropy/io/ascii/fastbasic.py in read(self, table)
    107                                       data_start=self.data_start,
    108                                       fill_extra_cols=self.fill_extra_cols,
--> 109                                       **self.kwargs)
    110         conversion_info = self._read_header()
    111         self.check_header()

~/Dropbox/Code/Astropy/astropy/astropy/io/ascii/cparser.pyx in astropy.io.ascii.cparser.CParser.__cinit__()

~/Dropbox/Code/Astropy/astropy/astropy/io/ascii/cparser.pyx in astropy.io.ascii.cparser.CParser.setup_tokenizer()

UnicodeEncodeError: 'ascii' codec can't encode character '\xeb' in position 11416: ordinal not in range(128)
```

However, pandas can read it in:

```python
In [6]: from pandas import read_csv

In [7]: read_csv('results.csv')
Out[7]: 
...
```

Don't ask why I'm trying to read in that table... 💃 🕺

**Comment by astrofrog** [comment]
This works though:

```
In [9]: t = Table.read('results.csv', format='ascii.csv', fast_reader=False)
```

But I thought with the guessing that not specifying the format should work as it should fall back to the non-fast version.

**Comment by bsipocz** [comment]
I remove the CSV mention from the title as this bug also affects other ascii tables (see https://www.facebook.com/groups/astropython/permalink/2031671610410852/?comment_id=2031688803742466&comment_tracking=%7B%22tn%22%3A%22R1%22%7D)

**Comment by saimn** [comment]
Hmm I don't know much about the guess logic but it works for me on master:
```
Python 3.6.3 (default, Dec  2 2017, 23:43:57) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %astropy
Numpy 1.13.3
Astropy 3.0.dev21140

In [2]: ascii.read('results.csv')
Out[2]: 
<Table length=1636>
celebrity_id professional_id       celebrity        ... score_donny score_shirley total
   int64          int64              str23          ...    int64        int64     int64
------------ --------------- ---------------------- ... ----------- ------------- -----
           0               0      Natasha Kaplinsky ...           0             0    27
           1               1         Lesley Garrett ...           0             0    29
           2               2           Chris Parker ...           0             0    19
           3               3             Jason Wood ...           0             0    21
           4               4          Verona Joseph ...           0             0    27
         ...             ...                    ... ...         ...           ...   ...
         200              34           Susan Calman ...           0             5    20
         197              47           Joe McFadden ...           0             8    32
         199              35        Aston Merrygold ...           0             9    35
         206              46        Alexandra Burke ...           0             8    33
         193              50         Davood Ghadami ...           0             7    25
```
It doesn't work if I specify the format (`ascii.read('results.csv', format='csv')`), unless I specify `fast_reader=False`.

**Comment by saimn** [comment]
Oh, it works because I use `ascii.read`, but I get the error with `Table.read` !

**Comment by taldcroft** [comment]
The difference is that `Table.read` goes through the unified I/O interface which auto-identifies the file format as `csv` based on the extension name.  This prevents the guessing from working.

So `Table.read('results.csv', format='ascii')` works fine.

There is still a bug that the fast `csv` reader is failing and the exception is not getting caught correctly.  In this case it should be falling through to the pure-Python `csv` format.  Sadly the logic of how the fast C reader and Python readers work together is pretty convoluted and prone to problems like this.

**Comment by taldcroft** [comment]
Here's a fix.  Anyone have time to put this in and write a test?
```
diff --git a/astropy/io/ascii/ui.py b/astropy/io/ascii/ui.py
index b494955..b3c1200 100644
--- a/astropy/io/ascii/ui.py
+++ b/astropy/io/ascii/ui.py
@@ -369,7 +369,7 @@ def read(table, guess=None, **kwargs):
                 _read_trace.append({'kwargs': fast_kwargs,
                                     'Reader': fast_reader_rdr.__class__,
                                     'status': 'Success with fast reader (no guessing)'})
-            except (core.ParameterError, cparser.CParserError) as err:
+            except (core.ParameterError, cparser.CParserError, UnicodeEncodeError) as err:
                 # special testing value to avoid falling back on the slow reader
                 if fast_reader['enable'] == 'force':
                     raise core.InconsistentTableError(
```
Maybe @vermashresth?

**Comment by vermashresth** [comment]
@taldcroft Sure, I will do!

**Comment by saimn** [comment]
Thanks for the explanation @taldcroft !
There is already a PR attempting to fix the issue : #7103

**Comment by taldcroft** [comment]
@saimn - duh, sorry not paying attention.  But at least the fix is the same (modulo a PEP8 error..)!
